internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -910,31 +987,185 @@ func (r *AgentServiceConfigReconciler) newImageServiceDeployment(ctx context.Con
-			deployment.Spec.Template.Spec.NodeSelector = r.NodeSelector
+			statefulSet.Spec.Template.Spec.NodeSelector = r.NodeSelector
 		} else {
-			deployment.Spec.Template.Spec.NodeSelector = map[string]string{}
+			statefulSet.Spec.Template.Spec.NodeSelector = map[string]string{}
 		}
 
 		if r.Tolerations != nil {
-			deployment.Spec.Template.Spec.Tolerations = r.Tolerations
+			statefulSet.Spec.Template.Spec.Tolerations = r.Tolerations
 		} else {
-			deployment.Spec.Template.Spec.Tolerations = []corev1.Toleration{}
+			statefulSet.Spec.Template.Spec.Tolerations = []corev1.Toleration{}
 		}
 		return nil
 	}
 
-	return deployment, mutateFn, nil
+	return statefulSet, mutateFn, nil
+}
+
+func (r *AgentServiceConfigReconciler) cleanupImageServiceFinalizer(ctx context.Context, statefulSet *appsv1.StatefulSet) error {
+	if !funk.ContainsString(statefulSet.GetFinalizers(), imageServiceStatefulSetFinalizerName) {
+		return nil
+	}
+
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	if err := r.List(ctx, pvcList, client.MatchingLabels{"app": imageServiceName}); err != nil {
+		return err
+	}
+
+	for i := range pvcList.Items {
+		if err := r.Client.Delete(ctx, &pvcList.Items[i]); err != nil {
+			return err
+		}
+	}
+
+	controllerutil.RemoveFinalizer(statefulSet, imageServiceStatefulSetFinalizerName)
+	if err := r.Update(ctx, statefulSet); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) reconcileImageServiceStatefulSet(ctx context.Context, log logrus.FieldLogger, instance *aiv1beta1.AgentServiceConfig) (*ctrl.Result, error) {
+	statefulSet, mutateFn, err := r.newImageServiceStatefulSet(ctx, log, instance)
+	if err != nil {
+		return &ctrl.Result{Requeue: true}, err
+	}
+
+	defer func() {
+		// delete old deployment if it exists and we've created the new stateful set correctly
+		// TODO: this can be removed when we no longer support upgrading from a release that used deployments for the image service
+		// NOTE: this relies on the err local variable being set correctly when the function returns
+		if err == nil {
+			_ = r.Client.Delete(ctx, &appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      imageServiceName,
+					Namespace: r.Namespace,
+				},
+			})
+		}
+	}()
+
+	key := client.ObjectKeyFromObject(statefulSet)
+	if err = r.Client.Get(ctx, key, statefulSet); err != nil {
+		// if the statefulset doesn't exist, create it
+		if !errors.IsNotFound(err) {
+			return &ctrl.Result{Requeue: true}, err
+		}
+		log.Info("Creating image service stateful set")
+		if err = mutateFn(); err != nil {
+			return &ctrl.Result{Requeue: true}, err
+		}
+		if err = r.Client.Create(ctx, statefulSet); err != nil {
+			return &ctrl.Result{Requeue: true}, err
+		}
+		return nil, nil
+	}
+
+	if !statefulSet.DeletionTimestamp.IsZero() {
+		if err = r.cleanupImageServiceFinalizer(ctx, statefulSet); err != nil {
+			return &ctrl.Result{Requeue: true}, err
+		}
+		return &ctrl.Result{Requeue: true}, nil
```

Comment:
Why force a requeue here?

Reply:
We could be in the case where someone changed the volume size.
That flow actually requires three loops through reconcile.
1. ASC is edited to change volume size
  - We reconcile by requesting the statefulset be deleted then requeue (your other comment)
2. Remove the finalizer
  - We reconcile here and requeue
3. Create the statefulset

It's very possible that we could get triggered by the side-effects of these actions, but I figured it was better to force a requeue rather than relying on something like the statefulset or PVs being removed to trigger a reconcile.

Does that make sense or do you think we should remove these?

Reply:
I think the forced requeue just invites the possibility that there will be an extra reconcile before deletion is complete and the cache has been updated as such. The delete event on the StatefulSet will cause a reconcile of the ACI to happen regardless, so I would just rely on that as part of the normal pattern:

1. Something about a watched resource changes
2. Reconcile runs and makes sure real-world state matches desired state.

Reply:
:+1: I'll test it out to be sure but if you think that both calling `Delete` on an object with a finalizer and also removing the finalizer will both definitely cause another reconcile then I can roll back this change.

The other nice thing this gave us was in the tests I could call reconcile until we didn't request a requeue, but I can just reconcile some set number of times to test these changes.